### PR TITLE
upgrade flux v2 dependencies take 4

### LINKED ISSUE #3878: Upgrade Flux v2 dependencies - Take 3
In order to fully support flux GA we need to upgrade weave gitops flux apis to use v2.0.0 apis

AC

- weave gitops using [flux v2.0.0 apis](https://github.com/fluxcd/flux2/releases/tag/v2.0.0)
- weave gitops should manage GitRepository, Kustomization and Receiver as v1 
- reduced scope of the impact for weave gitops ee by not upgrading controller runtime nor kube apis

### PR DESCRIPTION
**Pull request by chanwit**
Fixes #3878 

See the counter part #3880 which uses controller runtime v0.15

Notable differences:
```
// Replace k8s.io packages v0.26 to downgrade controller-runtime to v0.14.6
replace (
	k8s.io/api => k8s.io/api v0.26.3
	k8s.io/apiextensions-apiserver => k8s.io/apiextensions-apiserver v0.26.3
	k8s.io/apimachinery => k8s.io/apimachinery v0.26.3
	k8s.io/cli-runtime => k8s.io/cli-runtime v0.26.3
	k8s.io/client-go => k8s.io/client-go v0.26.3

	sigs.k8s.io/controller-runtime => sigs.k8s.io/controller-runtime v0.14.6
)
```

### CODE REVIEW

File: core/server/primarykinds.go [hunk 1]
```diff
@@ -18,6 +42,66 @@ func New() *PrimaryKinds {
...
+		return -1, nil
+	} else {
+		return 0, nil
+	}
+}
+
+// DefaultPrimaryKinds generates a new PrimaryKinds object which contains
+// the highest version of each kind of known types from a Kubernetes scheme.
+// It returns the PrimaryKinds object and any error that occurred.
 func DefaultPrimaryKinds() (*PrimaryKinds, error) {
```
Comment (enekofb) [inline core/server/primarykinds.go:105]: do you think that adding a test case for this method would help to provide understanding and regression?

**Review by enekofb — COMMENTED** [review]
fantastic work. 

some follow up topics before approving:
- verify `gitops run`
- understand experience for not v2 users (using with flux v0.4x) -> done [here](https://github.com/weaveworks/weave-gitops/pull/3889#issuecomment-1655281730)
- verify cli

and a couple of follow-up tickets
- to prepare release notes and update documentation https://docs.gitops.weave.works/docs/guides/fluxga-upgrade/
- to potentially implement multiple kind versions querying

**Comment by enekofb** [comment]
Understand experience for not v2 users (using with flux v0.4x)

Given wego with flux v0.41.2

Screenshot 2023-07-28 at 09 36 31

I could see errors for Sources for not finding v1/GitRepository

Screenshot 2023-07-28 at 09 37 21

I could see errors for Sources for not finding v1/Kustomizations

Screenshot 2023-07-28 at 09 37 56

File: go.mod [hunk 2]
```diff
@@ -10,21 +10,22 @@ require (
 	github.com/charmbracelet/lipgloss v0.6.0
 	github.com/cheshir/ttlcache v1.0.1-0.20220504185148-8ceeff21b789
 	github.com/coreos/go-oidc/v3 v3.1.0
-	github.com/fluxcd/flux2 v0.40.2
+	github.com/fluxcd/flux2/v2 v2.0.0-rc.3
```
[the remark this answers is not in the document]
Comment (chanwit) [reply go.mod:13]: `v2.0.0-rc.3` is correct. It does not affect the API modules anyway. This import is required by SSA lib, for example.
We cannot bump to v2.0.x. Doing so forces us to use Controller Runtime v0.15.

File: go.mod [hunk 3]
```diff
@@ -52,18 +53,18 @@ require (
...
-	k8s.io/cli-runtime v0.26.1
-	k8s.io/client-go v0.26.3
+	k8s.io/api v0.27.3
+	k8s.io/apiextensions-apiserver v0.27.3
+	k8s.io/apimachinery v0.27.3
+	k8s.io/cli-runtime v0.26.3
+	k8s.io/client-go v0.27.3
 	sigs.k8s.io/cli-utils v0.34.0
-	sigs.k8s.io/controller-runtime v0.14.6
+	sigs.k8s.io/controller-runtime v0.15.0
```
[the remark this answers is not in the document]
Comment (chanwit) [reply go.mod:67]: v0.15.0 is required by all API (like `source-controller/api`) modules by default. But it's downgradable as we do not refer to any specific v0.15.0 codes.

**Comment by enekofb** [comment]
quick test cli works in PR 

```
➜  weave-gitops git:(flux-v2-take-4-cr-14) ./bin/gitops create dashboard gitops --password=BDASDFASDF                                
✚ Generating GitOps Dashboard manifests ...
► Creating GitOps Dashboard objects ...
✚ Generating GitOps Dashboard manifests ...
✔ Generated GitOps Dashboard manifests
► Checking for a cluster in the kube config ...
► Checking if Flux is already installed ...
► Getting Flux version ...
✔ Flux v2.0.0 is already installed on the flux-system namespace.
► Applying GitOps Dashboard manifests
► Installing the GitOps Dashboard ...
✔ GitOps Dashboard has been installed
► Request reconciliation of dashboard (timeout 3m0s) ...
◎ Waiting for GitOps Dashboard reconciliation
✔ GitOps Dashboard gitops is ready
✔ Installed GitOps Dashboard

➜  weave-gitops git:(flux-v2-take-4-cr-14) ./bin/gitops check                                                                 <aws:sts>
✔ Kubernetes 1.24.0 >=1.20.6-0


Use "gitops remove [command] --help" for more information about a command.
➜  weave-gitops git:(flux-v2-take-4-cr-14) ./bin/gitops get config                                                            <aws:sts>
✔ Your CLI configuration for Weave GitOps:
{
  "analytics": true,
  "userId": "a5OoZqPd+p"
}
```

**Comment by chanwit** [comment]
`gitops run` confirmed to work
image

**Review by enekofb — APPROVED** [review]
🚀

**Pull request merged**